### PR TITLE
Add `arm64` support to .NET snap

### PR DESCRIPTION
**Pull request by mateusrodrigues**
This PR adds `arm64` support to the .NET snap.

- Adds `arm64` as a supported architecture in the `snapcraft.yaml` file and changes the build scripts accordingly.
- Moves the CI builds to the Canonical self-hosted machines so that `arm64` builds are also available.
- Makes minor change to the `Terminal` class by exposing the `InvocationResult` return type to the callers, which helped debug issues running the snap on `arm64`.

### CODE REVIEW

File: snap/snapcraft.yaml [hunk 1]
```diff
@@ -58,25 +65,34 @@ parts:
...
+      elif [ "${CRAFT_ARCH_BUILD_FOR}" = "arm64" ]; then
+        LIB_PATH=lib/aarch64-linux-gnu
+        INTERPRETER="/snap/dotnet/current/lib/ld-linux-aarch64.so.1"
+
+        patchelf --force-rpath --set-rpath \$ORIGIN/../${LIB_PATH}:\$ORIGIN/../../${LIB_PATH} \
+          ${CRAFT_STAGE}/usr/bin/jq
+        patchelf --force-rpath --set-rpath \$ORIGIN/${LIB_PATH}:\$ORIGIN/usr/${LIB_PATH} \
+          ${CRAFT_STAGE}/Dotnet.Installer.Console
+      else
+        echo "Unknown architecture (${CRAFT_ARCH_BUILD_FOR})"
```
Comment (dviererbe) [inline snap/snapcraft.yaml:90]: nitpick: The architecture is probably known, but unsupported.

```suggestion
        echo "Unsupported architecture (${CRAFT_ARCH_BUILD_FOR})"
```

File: snap/snapcraft.yaml [hunk 2]
```diff
@@ -86,10 +102,22 @@ parts:
...
       craftctl default
+
+      if [ "${CRAFT_ARCH_BUILD_FOR}" = "amd64" ]; then
+        INTERPRETER="/snap/dotnet/current/lib64/ld-linux-x86-64.so.2"
+        RPATH="/snap/dotnet/current/usr/lib/x86_64-linux-gnu"
+      elif [ "${CRAFT_ARCH_BUILD_FOR}" = "arm64" ]; then
+        INTERPRETER="/snap/dotnet/current/lib/ld-linux-aarch64.so.1"
+        RPATH="/snap/dotnet/current/usr/lib/aarch64-linux-gnu:/snap/dotnet/current/lib/aarch64-linux-gnu"
+      else
+        echo "Unknown architecture (${CRAFT_ARCH_BUILD_FOR})"
```
Comment (dviererbe) [inline snap/snapcraft.yaml:113]: nitpick: The architecture is probably known, but unsupported.

```suggestion
        echo "Unsupported architecture (${CRAFT_ARCH_BUILD_FOR})"
```

File: snap/snapcraft.yaml [hunk 3]
```diff
@@ -58,25 +65,34 @@ parts:
       fi
 
       /usr/bin/dotnet publish src/Dotnet.Installer.Console --output "${SNAPCRAFT_PART_INSTALL}" \
-        --configuration Release -r "${RUNTIME_RID}" -p:DebugSymbols=false -p:DebugType=none
+        --configuration Debug -r "${RUNTIME_RID}" -p:DebugSymbols=true -p:DebugType=full
```
Comment (dviererbe) [inline snap/snapcraft.yaml:68]: question: Why did you change `--configuration Release` to `--configuration Debug`?

File: .github/workflows/build-on-main.yaml [hunk 4]
```diff
@@ -85,14 +106,16 @@ jobs:
...
 
       - name: Install .NET
-        uses: actions/setup-dotnet@v4
-        with:
-          dotnet-version: '8.0.x'
+        id: install-dotnet
+        run: |
+          sudo apt-get update
+          sudo apt-get upgrade --yes
+          sudo apt-get install --yes dotnet-sdk-8.0
```
Comment (dviererbe) [inline .github/workflows/build-on-main.yaml:118]: praise: I like that we use our builds instead of the Microsoft builds :)

File: snap/snapcraft.yaml [hunk 5]
```diff
@@ -58,25 +65,34 @@ parts:
...
-      # .NET Installer files
-      patchelf --set-interpreter /snap/dotnet/current/lib64/ld-linux-x86-64.so.2 Dotnet.Installer.Console
-      patchelf --force-rpath --set-rpath \$ORIGIN/netcoredeps:/snap/dotnet/current/usr/lib/x86_64-linux-gnu Dotnet.Installer.Console
-    prime:
-      - usr/bin/jq
-      - usr/lib/*/libjq.so*
-      - usr/lib/*/libonig.so*
-      - Dotnet.Installer.Console
-      - Configuration/*
-      - Scripts/*
```
Comment (dviererbe) [inline snap/snapcraft.yaml:79]: question: Is it intended that the `prime` section is removed?

File: src/Dotnet.Installer.Core/Services/Implementations/SnapService.cs [hunk 6]
```diff
@@ -10,23 +10,21 @@ public bool IsSnapInstalled(string name, CancellationToken cancellationToken = d
...
         };
-        
+
         if (purge) arguments.Add("--purge");
         arguments.Add(name);
-        
-        var result = await Terminal.Invoke("snap", arguments.ToArray());
-        return new InvocationResult(result == 0, "", "");
+
+        return Terminal.Invoke("snap", arguments.ToArray());
```
Comment (dviererbe) [inline src/Dotnet.Installer.Core/Services/Implementations/SnapService.cs:28]: note: This was confusing to read, because I am used to the Async suffix of methods that return a Task or are marked `async` 😅

File: .github/workflows/build-on-main.yaml [hunk 7]
```diff
@@ -31,16 +39,25 @@ jobs:
...
           files: |
             TestResults/*.trx
 
   snapcraft-build:
-    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        os:
+          - [self-hosted, large, jammy, X64]
+          - [self-hosted, large, jammy, ARM64]
```
Comment (dviererbe) [inline .github/workflows/build-on-main.yaml:51]: question: How do we extend this eventually to `ppc64el` and `s390x`? There are no self-hosted GitHub runners for `ppc64el` and `s390x` as far as I am aware.

**Review by dviererbe — APPROVED** [review]
I left some questions and nitpicks. I approve this now, because I am on vacation for the next week.

File: snap/snapcraft.yaml (hunk 3 above)
Comment (mateusrodrigues) [reply snap/snapcraft.yaml:68]: Ugh, I committed this by accident 🤦🏼 I was debugging something and left this in there. Thanks for catching it!

File: snap/snapcraft.yaml (hunk 5 above)
Comment (mateusrodrigues) [reply snap/snapcraft.yaml:79]: Yes, the `stage` step will already contain all of the necessary files to prime, so no need to specify this in here.

I originally had a `craftctl default` line in the `override-build` section, which was copying everything from `build` to `stage`, requiring a `prime:` section later on. This is no longer the case.

File: .github/workflows/build-on-main.yaml (hunk 7 above)
Comment (mateusrodrigues) [reply .github/workflows/build-on-main.yaml:51]: Yes, there are no `s390x` and `ppc64el` self-hosted GitHub runners. I believe that we would need to move the builds to Launchpad/Snapcraft CI, but even then we would also need to move the content snap builds as well. It should be easy enough for this repo, but I'm not sure about the content snaps repo, since it contains multiple snap definitions in it.

File: src/Dotnet.Installer.Core/Services/Implementations/SnapService.cs (hunk 6 above)
Comment (mateusrodrigues) [reply src/Dotnet.Installer.Core/Services/Implementations/SnapService.cs:28]: Most of the methods in this project are either `async` or return `Task<T>`. I decided not to suffix them with `Async` because they would *mostly* be `*Async` 😆 

We can discuss this further and refactor the method names if necessary.

**Comment by mateusrodrigues** [comment]
@dviererbe thank you for the review. I'll leave this open for a few more days while I run some more tests on `arm64` machines, then I'll merge this.

**Comment by mateusrodrigues** [comment]
Tests complete. Merging this.
